[feature-wisun] Enable Nanostack DNS cache usage

artokin opened the pull request:
<!--
For more information on the requirements for pull requests, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html).

NOTE: Do not remove any of the template headings (even for optional sections) as this
template is automatically parsed. 
-->

### Summary of changes <!-- Required -->

<!-- 
    Please provide the following information: 

    Description of the the change (what is this fixing / adding / removing?).

    Why the change is needed (if this is fixing a reported issue please summarize what
    the issue is and add the reference. E.g. Fixes #17119).

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
    
-->
Enable usage of Nanostack DNS cache. The cache may contain DNS server address and DNS query results.

Inherit methods gethostbyname, gethostbyname_async and get_dns_server to Nanostack class. 
Methods will try to find DNS server address or DNS query results from Nanostack DNS cache.


#### Impact of changes <!-- Optional -->
<!-- 
    If there are any implications for users taking this change then they must be 
    provided here. For Major PR types this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

#### Migration actions required <!-- Optional -->
<!-- 
    This should only be applicable in Major PR types for which this field is MANDATORY.

    NOTE: This section is automatically written to release notes for Feature and 
    Major releases and should contain enough details for a user.
-->

### Documentation <!-- Required -->

<!-- 
    Please provide details of any document updates required, including links to any
    related PRs against the docs repository.
    If no document updates are required please specify 'None', this at least tells us
    that this has been considered.
-->

----------------------------------------------------------------------------------------------------------------
### Pull request type <!-- Required -->

<!--
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front of them would change the meaning incorrectly. 
-->
    [x] Patch update (Bug fix / Target update / Docs update / Test update / Refactor)
    [] Feature update (New feature / Functionality change / New API)
    [] Major update (Breaking change E.g. Return code change / API behaviour change)

----------------------------------------------------------------------------------------------------------------
### Test results <!-- Required -->

<!--
    Provide all the information required, listing all the testing performed. For new targets please attach full test results for all supported compilers.
-->
    [] No Tests required for this change (E.g docs only update)
    [x] Covered by existing mbed-os tests (Greentea or Unittest)
    [] Tests / results supplied as part of this PR
    
    
----------------------------------------------------------------------------------------------------------------
### Reviewers <!-- Optional -->

<!--
    Request additional reviewers with @username or @team
-->

----------------------------------------------------------------------------------------------------------------

Commit:
Mesh: Enable Nanostack DNS cache usage

Inherit methods gethostbyname, gethostbyname_async and get_dns_server
to Nanostack class. Methods will try to find DNS server address
or DNS query results from Nanostack DNS cache.

mikter commented on features/nanostack/nanostack-interface/Nanostack.cpp:
```diff
@@ -150,6 +158,50 @@ static int8_t find_interface_by_address(const uint8_t target_addr[16])
+        return -1;
+    }
+
+    interface_id = atoi(&interface_name[3]);
```
Do all interfaces have exactly 4 leters ? Mesh ETH? TUN? is there possibility to search first number and then atoi it instead of assuming

artokin replied:
I agree, this is problematic. Interface name is created https://github.com/ARMmbed/mbed-os/blob/feature-wisun/features/nanostack/mbed-mesh-api/source/MeshInterfaceNanostack.cpp#L132 and but it can also be given from application.

mikaleppanen replied:
Application returns interface name that the nanostack interfaces have returned, so nanostack is in full control of those. Nanostack interface naming pattern is to have always 3 letters and a number.

mikaleppanen commented on features/nanostack/nanostack-interface/Nanostack.cpp:
```diff
@@ -534,6 +586,74 @@ const char *Nanostack::get_ip_address()
+    if (nanostack_dns_query_result_check(name, address, interface_name) == 0) {
+        return 0;
+    }
+
```
Should have the "if (version == NSAPI_UNSPEC) { " and in that case set version ipv6. Resolving ipv4 addresses with nanostack is bit weird since stack does not have ipv4, but I guess no reason to force reject for those either.

mikaleppanen commented on features/nanostack/nanostack-interface/Nanostack.cpp:
```diff
@@ -534,6 +586,74 @@ const char *Nanostack::get_ip_address()
+
+    // try nanostack DNS cache, if not found then fallback to dns query
+    if (nanostack_dns_query_result_check(name, address, interface_name) == 0) {
+        return 0;
```
return NSAPI_ERROR_OK

mikaleppanen commented on features/nanostack/nanostack-interface/Nanostack.cpp:
```diff
@@ -534,6 +586,74 @@ const char *Nanostack::get_ip_address()
+        callback(NSAPI_ERROR_OK, &address);
+        return NSAPI_ERROR_OK;
+    }
+
```
also check for unspecified version

Commit:
Review corrections

-Check address version NSAPI_UNSPEC
-Disable traces
-More specific interface ID parsing
-Harmonize return values

Commit:
Follow Mbed OS coding style

Commit:
Update WisunBorderRouter class

Add set_dns_query_result to WisunBorderRouter.

Commit:
Follow Mbed OS coding style

Commit:
Use macro for isdigit
